src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp

```diff
@@ -124,30 +240,75 @@ namespace
+                    "cannot sample osl environment edf \"%s\" because it is not bound to the environment.",
+                    get_path().c_str());
+                value.set(0.0f);
+                probability = 0.0f;
+                return;
+            }
+
+            // Sample the importance map.
+            size_t x, y;
+            Color3f payload;
+            float prob_xy;
+            m_importance_sampler->sample(s, x, y, payload, prob_xy);
+            assert(prob_xy >= 0.0f);
+
+            // Compute the coordinates in [0,1)^2 of the sample.
+            const float jitter_x = frac(s[0] * m_importance_map_width);
+            const float jitter_y = frac(s[1] * m_importance_map_height);
+            const float u = (x + jitter_x) * m_rcp_importance_map_width;
+            const float v = (y + jitter_y) * m_rcp_importance_map_height;
+            assert(u >= 0.0f && u < 1.0f);
+            assert(v >= 0.0f && v < 1.0f);
+
+            // Compute the spherical coordinates of the sample.
+            float theta, phi;
+            unit_square_to_angles(u, v, theta, phi);
+            shift_angles(theta, phi, m_theta_shift, m_phi_shift);
+
+            // Compute the local space emission direction.
+            const float cos_theta = cos(theta);
+            const float sin_theta = sin(theta);
+            const float cos_phi = cos(phi);
+            const float sin_phi = sin(phi);
+            const Vector3f local_outgoing =
+                Vector3f::make_unit_vector(cos_theta, sin_theta, cos_phi, sin_phi);
+
+            // Transform the emission direction to world space.
             Transformd scratch;
             const Transformd& transform = m_transform_sequence.evaluate(0.0f, scratch);
             outgoing = transform.vector_to_parent(local_outgoing);
 
-            evaluate_osl_background(shading_context, local_outgoing, value);
+            // Return the emitted radiance.
+            value.set(payload, g_std_lighting_conditions, Spectrum::Illuminance);
+
+            // Compute the probability density of this direction.
+            probability = prob_xy * m_probability_scale / sin_theta;
+            assert(probability >= 0.0f);            
         }
 
+        // Calculates only value, therefore importance map is not needed.
         void evaluate(
             const ShadingContext&   shading_context,
             const Vector3f&         outgoing,
             Spectrum&               value) const override
         {
             assert(is_normalized(outgoing));
 
+            // Transform the emission direction to local space.
             Transformd scratch;
             const Transformd& transform = m_transform_sequence.evaluate(0.0f, scratch);
             const Vector3f local_outgoing = transform.vector_to_local(outgoing);
 
             evaluate_osl_background(shading_context, local_outgoing, value);
         }
 
+        // Calculates value and probability, therefore importance map is needed.
```

Comment:
This comment would be best placed right before the line

    if (!m_importance_sampler)